src/components/ReviewForm.js

```diff
@@ -0,0 +1,176 @@
+import React, { useState } from "react";
+import axios from "axios";
+
+export function ReviewForm(props) {
+  const { id } = props.editableReview;
+  const [errorTitleMessage, setErrorTitleMessage] = useState("");
+  const [errorScoreMessage, setErrorScoreMessage] = useState("");
+  const [errorBodyMessage, setErrorBodyMessage] = useState("");
+  const [errorReviewerMessage, setErrorReviewerMessage] = useState("");
+  const [formData, setFormData] = useState(
+    id
+      ? props.editableReview
+      : {
+          title: "",
+          score: 1,
+          body: "",
+          reviewer: "",
+        }
+  );
+  const actionText = id ? "レビューを編集する" : "レビューを登録する";
+
+  function handleChange({ target }) {
+    const value = target.value;
+    console.log(value);
+    setFormData({
+      ...formData,
+      [target.name]: value,
+    });
+  }
+
+  // async関数にしてあげる　updateReviewをawaitするために
+  async function handleSubmit(event) {
+    event.preventDefault(); // 画面遷移してしまわないように必要
+    console.log(formData);
+    if (validate(formData)) {
+      if (id) {
+        await updateReview(id);
+      } else {
+        postReview();
+      }
+    }
+    props.closeModal();
+  }
+
+  function validate({ title, score, body, reviewer }) {
+    if (title && title.length <= 255) {
+      setErrorTitleMessage("");
+    } else {
+      setErrorTitleMessage("Title is invalid");
+    }
+
+    if (score && 0 < score <= 5) {
+      setErrorScoreMessage("");
+    } else {
+      setErrorScoreMessage("Score is invalid");
+    }
+
+    if (body && body.length <= 255) {
+      setErrorBodyMessage("");
+    } else {
+      setErrorBodyMessage("Body is invalid");
+    }
+
+    if (reviewer && reviewer.length <= 255) {
+      setErrorReviewerMessage("");
+    } else {
+      setErrorReviewerMessage("Reviewer is invalid");
+    }
+
+    if (
+      !(
+        errorTitleMessage &&
+        errorScoreMessage &&
+        setErrorBodyMessage &&
+        errorReviewerMessage
+      )
+    ) {
+      return true;
+    } else {
+      return false;
+    }
+  }
```

Comment:
`setXXX` は呼ばれた直後は state が前の state のままとなり期待どおりの挙動をしないのでこの関数は以下のように書き直した方がよさそうです

```suggestion
  function validate({ title, score, body, reviewer }) {
    let valid = true;

    if (title && title.length <= 255) {
      setErrorTitleMessage("");
    } else {
      valid = false;
      setErrorTitleMessage("Title is invalid");
    }

    if (score && 0 < score && score <= 5) {
      setErrorScoreMessage("");
    } else {
      valid = false;
      setErrorScoreMessage("Score is invalid");
    }

    if (body && body.length <= 255) {
      setErrorBodyMessage("");
    } else {
      valid = false;
      setErrorBodyMessage("Body is invalid");
    }

    if (reviewer && reviewer.length <= 255) {
      setErrorReviewerMessage("");
    } else {
      valid = false;
      setErrorReviewerMessage("Reviewer is invalid");
    }

    return valid;
  }
```

see: https://medium.com/ableneo/react-setstate-does-not-immediately-update-the-state-84dbd26f67d5

Reply:
createの時だけvalidationが動かなくて困っていました。解決しました✨